funsor/tensor_ops.py

```diff
@@ -0,0 +1,149 @@
+
+    This is mainly useful for implementing eager funsor operations.
+
+    :param funsor.terms.Funsor \*args: Multiple :class:`Tensor` s or :class:`Array` s and
+        :class:`~funsor.terms.Number` s.
+    :param bool expand: Whether to expand input tensors. Defaults to False.
+    :return: a pair ``(inputs, tensors)`` where tensors are all
+        :class:`torch.Tensor` s that can be broadcast together to a single data
+        with given ``inputs``.
+    :rtype: tuple
+    """
+    expand = kwargs.pop('expand', False)
+    assert not kwargs
+    inputs = OrderedDict()
+    for x in args:
+        inputs.update(x.inputs)
+    tensors = [align_tensor(inputs, x, expand=expand) for x in args]
+    return inputs, tensors
+
+
+def arange(prototype, name, *args, **kwargs):
+    """
+    Helper to create a named :func:`torch.arange` funsor.
+    In some cases this can be replaced by a symbolic
+    :class:`~funsor.terms.Slice` .
+
+    :param prototype: either a torch.Tensor or a numpy.array
+    :param str name: A variable name.
+    :param int start:
+    :param int stop:
+    :param int step: Three args following :py:class:`slice` semantics.
+    :param int dtype: An optional bounded integer type of this slice.
+    :rtype: Tensor
+    """
+    start = 0
+    step = 1
+    dtype = None
+    if len(args) == 1:
+        stop = args[0]
+        dtype = kwargs.pop("dtype", stop)
+    elif len(args) == 2:
+        start, stop = args
+        dtype = kwargs.pop("dtype", stop)
+    elif len(args) == 3:
+        start, stop, step = args
+        dtype = kwargs.pop("dtype", stop)
+    elif len(args) == 4:
+        start, stop, step, dtype = args
+    else:
+        raise ValueError
+    if step <= 0:
+        raise ValueError
+    stop = min(dtype, max(start, stop))
+    data = ops.new_arange(prototype, start, stop, step)
+    inputs = OrderedDict([(name, bint(len(data)))])
+    return Tensor(data, inputs, dtype=dtype)
+
+
+def materialize(prototype, x):
+    """
+    Attempt to convert a Funsor to a :class:`~funsor.terms.Number` or
+    :class:`Tensor` by substituting :func:`arange` s into its free variables.
+
+    :param prototype: either a torch.Tensor or a numpy.array
+    :arg Funsor x: A funsor.
+    :rtype: Funsor
+    """
+    assert isinstance(x, Funsor)
+    if isinstance(x, (Number, funsor.torch.Tensor, funsor.numpy.Array)):
+        return x
+    subs = []
+    for name, domain in x.inputs.items():
+        if isinstance(domain.dtype, int):
+            subs.append((name, arange(prototype, name, domain.dtype)))
+    subs = tuple(subs)
+    return substitute(x, subs)
+
+
+@dispatch(torch.Tensor, (type(None), tuple, OrderedDict), (int, str))
+def _Tensor(x, inputs, dtype):
+    return funsor.torch.Tensor(x, inputs, dtype)
+
+
+@dispatch(np.ndarray, (type(None), tuple, OrderedDict), (int, str))
+def _Tensor(x, inputs, dtype):
+    return funsor.numpy.Array(x, inputs, dtype)
+
+
+def Tensor(x, inputs=None, dtype="real"):
```

Comment:
This is very confusing and should be renamed.

Reply:
Initially, I wanted to change torch.Tensor to torch.TorchTensor so less code will be needed to change. But I have less opinion on naming, so I would like to incorporate your suggestion. How about NumericArray? And we use NumericArray throughout the repo...